dali/pipeline/operators/reader/reader_op.h

```diff
@@ -49,58 +49,71 @@ class DataReader : public Operator<Backend> {
  public:
   inline explicit DataReader(const OpSpec& spec) :
     Operator<Backend>(spec),
-  prefetched_batch_ready_(false),
   prefetch_ready_workers_(false),
   finished_(false),
+  prefetch_queue_depth_(spec.GetArgument<int>("prefeth_queue_depth")),
+  prefetched_batch_queue_(prefetch_queue_depth_),
+  curr_batch_consumer_(0),
+  curr_batch_producer_(0),
+  consumer_cycle_(false),
+  producer_cycle_(false),
   samples_processed_(0),
   batch_stop_(false),
-  prefetch_error_(nullptr) {
-  }
+  prefetch_error_(nullptr) {}
 
   ~DataReader() noexcept override {
     StopPrefetchThread();
-    for (size_t i = 0; i < prefetched_batch_.size(); ++i) {
-      // return unconsumed batches
-      if (prefetched_batch_[i]) {
-        loader_->ReturnTensor(prefetched_batch_[i]);
+    for (auto p_b : prefetched_batch_queue_) {
+      for (size_t i = 0; i < p_b.size(); ++i) {
+        // return unconsumed batches
+        if (p_b[i]) {
+          loader_->ReturnTensor(p_b[i]);
+        }
       }
     }
   }
 
   // perform the prefetching operation
   virtual void Prefetch() {
-    prefetched_batch_.reserve(Operator<Backend>::batch_size_);
-    prefetched_batch_.clear();
+    // We actually prepare the next batch
+    TimeRange tr("Prefetching for " + to_string(curr_batch_producer_), TimeRange::kRed);
+    prefetched_batch_queue_[curr_batch_producer_].reserve(Operator<Backend>::batch_size_);
+    prefetched_batch_queue_[curr_batch_producer_].clear();
 
     for (int i = 0; i < Operator<Backend>::batch_size_; ++i) {
       auto* t = loader_->ReadOne();
-      prefetched_batch_.push_back(t);
+      prefetched_batch_queue_[curr_batch_producer_].push_back(t);
     }
   }
 
   // Main prefetch work loop
   void PrefetchWorker() {
-    std::unique_lock<std::mutex> lock(prefetch_access_mutex_);
-
-    // if a result is already ready, wait until it's consumed
-    producer_.wait(lock, [&]() { return !prefetched_batch_ready_; });
+    {
+      std::unique_lock<std::mutex> lock(prefetch_access_mutex_);
 
+      // if a result is already ready, wait until it's consumed
+      producer_.wait(lock, [&]() { return !IsFull(); });
+    }
     while (!finished_) {
       try {
         Prefetch();
       } catch (const std::exception& e) {
         prefetch_error_.reset(new std::string(e.what()));
-        prefetched_batch_ready_ = true;
+        finished_ = true;
```

Comment:
This should be also protected by lock(prefetch_access_mutex_);

Reply:
please check after lasts commit